stdlib/mexpr/parser.mc

```diff
@@ -0,0 +1,433 @@
+  sem parseExprImp (p: Pos) =
+  | "if" ++ xs ->
+     let e1 = parseExprMain (advanceCol p 2) 0 xs in
+     let r1 = matchKeyword "then" e1.pos e1.str  in
+     let e2 = parseExprMain r1.pos 0 r1.str in
+     let r2 = matchKeyword "else" e2.pos e2.str  in
+     let e3 = parseExprMain r2.pos 0 r2.str in
+     {val = TmMatch {target = e1.val, pat = PBool {val = true, fi = NoInfo ()},
+                     thn = e2.val, els = e3.val, fi = makeInfo p e3.pos},
+      pos = e3.pos,
+      str = e3.str}
+ end
+
+-- Integer arithmetic parser (intrinsics)
+lang ArithIntParser = ExprParser + ConstAst + ArithIntAst
+  sem parseExprImp (p: Pos) =
+  | "addi" ++ xs -> let p2 = advanceCol p 4 in
+      {val = TmConst {val = CAddi {}, fi = makeInfo p p2}, pos = p2, str = xs}
+  | "subi" ++ xs -> let p2 = advanceCol p 4 in
+      {val = TmConst {val = CSubi {}, fi = makeInfo p p2}, pos = p2, str = xs}
+  | "muli" ++ xs -> let p2 = advanceCol p 4 in
+      {val = TmConst {val = CMuli {}, fi = makeInfo p p2}, pos = p2, str = xs}
+end
+
+
+-- Parse parentheses
+lang ParenthesesParser = ExprParser + KeywordParser
+  sem parseExprImp (p: Pos) =
+  | "(" ++ xs ->
+    let e = parseExprMain (advanceCol p 1) 0 xs in
+    let r = matchKeyword ")" e.pos e.str in
+    {val = e.val, pos = r.pos, str = r.str}
+end
+
+-- Parses a sequence of 
+lang SeqParser = ExprParser + KeywordParser + SeqAst
+  sem parseExprImp (p: Pos) =
+  | "[" ++ xs -> 
+      recursive let work = lam acc. lam first. lam p2. lam str.
+        let r = eatWSAC p2 str in
+  	match r.str with "]" ++ xs then
+	  {val = TmSeq{tms = acc, fi = makeInfo p (advanceCol r.pos 1)},
+	   pos = advanceCol r.pos 1, str = xs}
+	else
+	  let r2 = if first then r else matchKeyword "," r.pos r.str in
+	  let e = parseExprMain r2.pos 0 r2.str in
+	  work (snoc acc e.val) false e.pos e.str  
+      in work [] true (advanceCol p 1) xs
+end
+
+-- Sequence operations (intrinsics)
+lang SeqOpParser = ExprParser + ConstAst + SeqOpAst
+  sem parseExprImp (p: Pos) =
+  | "concat" ++ xs -> let p2 = advanceCol p 6 in
+      {val = TmConst {val = CConcat {}, fi = makeInfo p p2}, pos = p2, str = xs}
+end
+
+-- Matches a character (including escape character). 
+let matchChar : Pos -> String -> {val: Char, pos: Pos, str: String} =
+  lam p. lam str.
+  let ret = lam c. lam s. lam n. {val = c, pos = (advanceCol p n), str = s} in
+    match str with "\\" ++ xs then
+      match xs with "\\" ++ xs then ret '\\' xs 2 else
+      match xs with "n" ++ xs then ret '\n' xs 2 else
+      match xs with "t" ++ xs then ret '\t' xs 2 else
+      match xs with "\"" ++ xs then ret '\"' xs 2 else 
+      match xs with "'" ++ xs then ret '\'' xs 2 else
+      posErrorExit (advanceCol p 1) "Unknown escape character."
+    else match str with [x] ++ xs then ret x xs 1
+    else posErrorExit p "Unexpected end of file."
+    -- TODO David (2020-09-27): Shoud we allow newlines etc. inside strings
```

Comment:
I believe the todo style we settled on was not quite this (`TODO(David, 2020-09-27): ...`)

Reply:
Good. I have updated to the style we decided on at the meeting.